app/services/stripe_model_callbacks/invoice/updated_service.rb

```diff
@@ -1,23 +1,40 @@
 class StripeModelCallbacks::Invoice::UpdatedService < StripeModelCallbacks::BaseEventService
+  TRACKED_ACTIVITIES = {
+    "invoice.payment_failed": :payment_failed,
+    "invoice.payment_succeeded": :payment_succeeded,
+    "invoice.sent": :sent,
+    "invoice.upcoming": :upcoming
+  }.freeze.with_indifferent_access
+  private_constant :TRACKED_ACTIVITIES
+
   def execute
     invoice.assign_from_stripe(object)
+    return success_actions if invoice.save
 
-    if invoice.save
-      create_activity
+    fail!(invoice.errors.full_messages)
+  rescue ActiveRecord::RecordNotUnique, PG::UniqueViolation, SQLite3::ConstraintException => e
```

Comment:
This stings a bit. So the problem is that two events for the same Stripe object can be handled at the same time, right?

This problem could happen for pretty much any model and not just the invoice one, right?

Couldn't we use a lock to make sure, that only one operation per Stripe ID is being handled simultaneously? 

I am thinking about using this gem: https://github.com/ClosureTree/with_advisory_lock

```ruby
StripeModelCallbacks::ApplicationRecord.with_advisory_lock("STRIPE-OBJECT-#{object.id}") do
  # Do what needs to be done.
end
```

Reply:
@kaspernj this sounds good!